packages/spacecat-shared-rum-api-client/src/functions/cwv.js

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { quantile } from 'd3-array';
+import { pageviewsByUrl } from '../common/aggregateFns.js';
+import { FlatBundle } from '../common/flat-bundle.js';
+
+const CWV_METRICS = ['lcp', 'cls', 'inp', 'ttfb'].map((metric) => `cwv-${metric}`);
+
+function collectCWVs(groupedByUrlAndSource) {
+  const { url, items: itemsByUrl } = groupedByUrlAndSource;
+
+  const CWVs = itemsByUrl.reduce((acc, { items }) => {
+    // each session (id) can contain multiple measurement for the same metric
+    // we need to find the maximum per metric type
+    const maximums = items.reduce((values, item) => {
+      // eslint-disable-next-line no-param-reassign
+      values[item.checkpoint] = Math.max(values[item.checkpoint] || 0, item.value);
+      return values;
+    }, {});
+
+    // max values per id for each metric type are collected into an array
+    CWV_METRICS.forEach((metric) => {
+      if (!acc[metric]) acc[metric] = [];
+      if (maximums[metric]) {
+        acc[metric].push(maximums[metric]);
+      }
+    });
+    return acc;
+  }, {});
+
+  return {
+    url,
+    lcp: quantile(CWVs['cwv-lcp'], 0.75) || null,
+    lcpCount: CWVs['cwv-lcp'].length,
+    cls: quantile(CWVs['cwv-cls'], 0.75) || null,
+    clsCount: CWVs['cwv-cls'].length,
+    inp: quantile(CWVs['cwv-inp'], 0.75) || null,
+    inpCount: CWVs['cwv-inp'].length,
+    ttfb: quantile(CWVs['cwv-ttfb'], 0.75) || null,
+    ttfbCount: CWVs['cwv-ttfb'].length,
```

Comment:
Is this 75th percentile by date or across all dates spanning the given interval?

I ask because `id` is not guaranteed to be unique across dates.  The preferred grouping logic for events is unique combination of id+url+date.  Line 31 comment looks appropriate for a single day, but the quantile calls look appropriate for an interval of days.

Reply:
incorporated: https://github.com/adobe/spacecat-shared/pull/252/commits/53e2fa01d932163d66e7a299381c6241ca99bfd5